src/strnum.F90

```diff
@@ -0,0 +1,73 @@
+!> @file
+!> @brief Decode an integer from a character string.
+!>
+!> @author J. Woollen @date 1994-01-06
+
+!> This subroutine decodes an integer from a character string.  The
+!> string may contain leading or trailing blanks, but otherwise should
+!> contain only digits and an (optional) leading sign ('+' or '-')
+!> character. 
+!>
+!> If the string contains all blank characters, then num is returned
+!> with a value of 0.
+!>
+!> @param[in]  str -- character*(*): String
+!> @param[out] num -- integer: Value decoded from str
+!> @param[out] iret -- return code:
+!>   0 = success
+!>  -1 = string contained one or more illegal characters
+!>
+!> @author J. Woollen @date 1994-01-06
+recursive subroutine strnum(str,num,iret)
+  use modv_im8b
+
+  implicit none
+
+  character*(*), intent(in) :: str
+
+  integer, intent(out) ::  num, iret
+
+  character str2*40, fmt*8
+
+  integer lens
+
+!-----------------------------------------------------------------------
+!-----------------------------------------------------------------------
+
+! Check for I8 integers.
+
+  if (im8b) then
+     im8b = .false.
+
+     call strnum ( str, num, iret )
+     call x48 ( num, num, 1 )
+     call x48 ( iret, iret, 1 )
+
+     im8b = .true.
+     return
+  end if
+
+  iret = 0
+
+! Check for blank input string.
+
+  if ( str .eq. ' ' ) then
```

Comment:
What happens if str is '  '? (That is, two spaces, not one?) 

Is there a test that tries this function with different bad inputs to see if it catches them?

Until now, we have been relying on intestN and outtestN to test lots of functions. These are more like system tests than unit tests. What would be good would be (in addition) a new test, something like test_strings.F90, which can test all the string handling functions of NCEPLIBS-bufr. In that new test, you can systematically check the strnum subprogram with a variety of inputs and check that it catches all error conditions. See test_misc.F90 for an example of a test that runs these sorts of simple unit tests.

Reply:
This check will work for a string of any number of blanks (not just a single blank).  Within this PR I've already modified intest7.F90 accordingly to confirm this (it tests for a string of 4 blanks ;-).

That said, if you'd prefer I/we can move these new strnum tests over from intest7.F90 to test_misc.F90 - just let me know.